WalletWasabi/WabiSabi/Models/MultipartyTransaction/MultipartyTransactionState.cs

```diff
@@ -13,19 +20,44 @@ protected MultipartyTransactionState(MultipartyTransactionParameters parameters)
 
 	public MultipartyTransactionParameters Parameters { get; }
 
-	public ImmutableList<Coin> Inputs { get; init; } = ImmutableList<Coin>.Empty;
-	public ImmutableList<TxOut> Outputs { get; init; } = ImmutableList<TxOut>.Empty;
+	[JsonIgnore]
+	public IEnumerable<Coin> Inputs => Events.OfType<InputAdded>().Select(x => x.Coin);
+	[JsonIgnore]
+	public IEnumerable<TxOut> Outputs => Events.OfType<OutputAdded>().Select(x => x.Output);
 
+	[JsonIgnore]
 	public Money Balance => Inputs.Sum(x => x.Amount) - Outputs.Sum(x => x.Value);
+	[JsonIgnore]
 	public int EstimatedInputsVsize => Inputs.Sum(x => x.TxOut.ScriptPubKey.EstimateInputVsize());
+	[JsonIgnore]
 	public int OutputsVsize => Outputs.Sum(x => x.ScriptPubKey.EstimateOutputVsize());
 
+	[JsonIgnore]
 	public int EstimatedVsize => MultipartyTransactionParameters.SharedOverhead + EstimatedInputsVsize + OutputsVsize;
+	[JsonIgnore]
 	public int MaxTransactionSize => Parameters.MaxTransactionSize;
 
 	// With no coordinator fees we can't ensure that the shared overhead
 	// of the transaction also pays at the nominal feerate so this will have
 	// to do for now, but in the future EstimatedVsize should be used
 	// including the shared overhead
+	[JsonIgnore]
 	public FeeRate EffectiveFeeRate => new(Balance, EstimatedInputsVsize + OutputsVsize);
+
+	public ImmutableList<IEvent> Events { get; init; } = ImmutableList<IEvent>.Empty;
+
+	public MultipartyTransactionState GetConstructionStateSince(int order) =>
+		this with {
+			Events = Events.Skip(order).ToImmutableList()
+		};
+
+	public MultipartyTransactionState Merge(MultipartyTransactionState diff) =>
+		this with {
+			Events = Events.AddRange(diff.Events)
+		};
+
+	public MultipartyTransactionState MergeBack(MultipartyTransactionState origin) =>
```

Comment:
I am having a hard time understanding this. Does it seem like MergeFrom?

Reply:
yes, it sucks. I will improve it.

Reply:
It has to be better now.